### PR TITLE
Add historic metrics

### LINKED ISSUE #326: Update metrics collector to store historic app stats
Metric collector has to be updated to store historic data about apps:

- DB should be added
- Probably should be merged into stats-api repo
- In the first version we need to store data about transactions from the app for:

1. last day
2. last 7 days
3. last 30 days
4. total txs

More details will be added later.

### PR DESCRIPTION
**Pull request by dmytrotkk**
In this PR:

- Added historic metrics collection
- Added MySQL DB to store historic metrics
- Added new fields in metrics.json: last day tx, last 7 days tx, last 30 days tx
- Added types for metrics data
- Added linter and tests pipeline for metrics

### CODE REVIEW

File: .github/workflows/test_metrics.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  test_proxy:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.12]
+    env:
+      ETH_ENDPOINT: ${{ secrets.ETH_ENDPOINT }}
+      PYTHONPATH: ${{ github.workspace }}/metrics
+    steps:
+    - uses: actions/checkout@v2
```
Comment (DimaStebaev) [inline .github/workflows/test_metrics.yml]: v4 is here

File: .github/workflows/test_metrics.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      with:
+        submodules: true
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v1
```
Comment (DimaStebaev) [inline .github/workflows/test_metrics.yml]: v5 is here

File: .github/workflows/test_metrics.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      with:
```
Comment (DimaStebaev) [inline .github/workflows/test_metrics.yml:22]: Add
`cache: 'pip'`

File: metrics/src/collector.py [hunk 4]
```diff
@@ -21,93 +21,162 @@
...
 
 
-def get_metadata_url(network_name: str):
-    return f'https://raw.githubusercontent.com/skalenetwork/skale-network/master/metadata/{network_name}/chains.json' # noqa
+def get_metadata_url(network_name: str) -> str:
+    return f'{GITHUB_RAW_URL}/skalenetwork/skale-network/master/metadata/{network_name}/chains.json'
 
 
-def download_metadata(network_name: str):
+async def download_metadata(session, network_name: str):
```
Comment (DimaStebaev) [inline metrics/src/collector.py]: Return type?

File: metrics/src/config.py [hunk 5]
```diff
@@ -19,36 +19,41 @@
 
 import os
 
-ERROR_TIMEOUT = os.getenv('MONITOR_INTERVAL', 60)
-MONITOR_INTERVAL = os.getenv('MONITOR_INTERVAL', 10800)
-NETWORK_NAME = os.getenv('NETWORK_NAME', 'mainnet') or 'mainnet'
-ENDPOINT = os.environ['ETH_ENDPOINT']
+ERROR_TIMEOUT = os.getenv("MONITOR_INTERVAL", 60)
```
Comment (DimaStebaev) [inline metrics/src/config.py]: Why double quotes?

File: metrics/src/db.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+
+async def get_current_total_transactions(session, chain_name: str, address: str) -> int:
+    url = get_address_counters_url(NETWORK_NAME, chain_name, address)
+    async with session.get(url) as response:
+        data = await response.json()
+        return int(data.get('transactions_count', 0))
+
+
+async def bootstrap_db(session, apps_data: Dict[str, Dict[str, List[str]]]):
```
Comment (DimaStebaev) [inline metrics/src/db.py]: Return type?

File: metrics/src/main.py [hunk 7]
```diff
@@ -18,31 +18,115 @@
...
+            logger.info(
+                f'Not time for collection yet. \
+                Last run: {last_run_date}, Current date: {current_date}'
+            )
+            logger.info(f'Sleeping for {METRICS_CHECK_INTERVAL} seconds.')
+            sleep(METRICS_CHECK_INTERVAL)
+
+
+def wait_for_db():
+    max_retries = 30
```
Comment (DimaStebaev) [inline metrics/src/main.py]: Move to the config?

File: metrics/src/main.py [hunk 8]
```diff
@@ -18,31 +18,115 @@
...
+
+async def bootstrap_database():
+    logger.info('Checking if database needs bootstrapping...')
+    try:
+        async with aiohttp.ClientSession() as session:
+            metadata = await download_metadata(session, NETWORK_NAME)
+
+            apps_data = {}
+            for chain_name, chain_info in metadata.items():
+                if chain_name != '__offchain' and 'apps' in chain_info:
```
Comment (DimaStebaev) [inline metrics/src/main.py]: Read the literal from config.

File: scripts/backup_db.sh [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
```
Comment (DimaStebaev) [inline scripts/backup_db.sh:3]: `set -e` ?

File: scripts/restore_db.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
```
Comment (DimaStebaev) [inline scripts/restore_db.sh:2]: `set -e`?

File: .github/workflows/test_metrics.yml (hunk 1 above)
Comment (dmytrotkk) [reply .github/workflows/test_metrics.yml]: updated to v4

File: .github/workflows/test_metrics.yml (hunk 2 above)
Comment (dmytrotkk) [reply .github/workflows/test_metrics.yml]: updated to v5

File: .github/workflows/test_metrics.yml (hunk 3 above)
Comment (dmytrotkk) [reply .github/workflows/test_metrics.yml:22]: added

File: metrics/src/collector.py (hunk 4 above)
Comment (dmytrotkk) [reply metrics/src/collector.py]: added generic type

File: metrics/src/config.py (hunk 5 above)
Comment (dmytrotkk) [reply metrics/src/config.py]: fixed

File: metrics/src/db.py (hunk 6 above)
Comment (dmytrotkk) [reply metrics/src/db.py]: added

File: metrics/src/main.py (hunk 7 above)
Comment (dmytrotkk) [reply metrics/src/main.py]: moved to config

File: metrics/src/main.py (hunk 8 above)
Comment (dmytrotkk) [reply metrics/src/main.py]: done

File: scripts/backup_db.sh (hunk 9 above)
Comment (dmytrotkk) [reply scripts/backup_db.sh:3]: added

File: scripts/restore_db.sh (hunk 10 above)
Comment (dmytrotkk) [reply scripts/restore_db.sh:2]: added

**Review by badrogger — APPROVED** [review]

File: metrics/src/db.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+from src.config import NETWORK_NAME
+from src.explorer import get_address_counters_url
+
+logger = logging.getLogger(__name__)
+
+TRANSACTION_COUNT_FIELD = 'transactions_count'
+BACKFILL_DB_DAYS = 30
+
+
+async def get_current_total_transactions(session, chain_name: str, address: str) -> int:
```
Comment (DmytroNazarenko) [inline metrics/src/db.py]: I think it is better to move this method into src.explorer

File: metrics/src/gas.py [hunk 12]
```diff
@@ -19,33 +19,33 @@
...
 from web3 import Web3
-from config import ENDPOINT, GAS_ESTIMATION_ITERATIONS, BLOCK_SAMPLING
+from src.config import ENDPOINT, GAS_ESTIMATION_ITERATIONS, BLOCK_SAMPLING
 
 logger = logging.getLogger(__name__)
 
 
 def init_w3():
-    logger.info(f'Connecting to {ENDPOINT}...')
+    logger.info(f"Connecting to {ENDPOINT}...")
```
Comment (DmytroNazarenko) [inline metrics/src/gas.py]: Why double-quotes, but not single-quotes in this module?

File: metrics/src/main.py [hunk 13]
```diff
@@ -18,31 +18,117 @@
...
+from src.models import db
+from src.db import bootstrap_db
 
 logger = logging.getLogger(__name__)
 
 
+def run_migrations():
+    logger.info('Running database migrations...')
+    try:
+        from models import Address, TransactionCount
```
Comment (DmytroNazarenko) [inline metrics/src/main.py]: Move models import to the other imports

File: metrics/src/main.py [hunk 14]
```diff
@@ -18,31 +18,117 @@
...
         sys.exit(1)
+
+    logger.info(f'Starting metrics collection loop for network: {NETWORK_NAME}')
+    last_run_date = None
+
     while True:
-        logger.info('Metrics collector iteration started...')
+        current_date = datetime.now().date()
+
+        if last_run_date is None or current_date > last_run_date:
```
Comment (DmytroNazarenko) [inline metrics/src/main.py:66]: If you are running `collect_metrics` at the beginning of the day, then in `daily_transactions` field for this date in db will be in fact daily transactions count from the previous day. Is it supposed to work like this?

File: metrics/src/db.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+from peewee import fn, IntegrityError
+
+from src.models import db, Address, TransactionCount
+from src.config import NETWORK_NAME
+from src.explorer import get_address_counters_url
+
+logger = logging.getLogger(__name__)
+
+TRANSACTION_COUNT_FIELD = 'transactions_count'
+BACKFILL_DB_DAYS = 30
```
Comment (DmytroNazarenko) [inline metrics/src/db.py]: Probably it's better to move db constants to src/config.py

File: metrics/src/db.py (hunk 11 above)
Comment (dmytrotkk) [reply metrics/src/db.py]: moved

File: metrics/src/main.py (hunk 14 above)
Comment (dmytrotkk) [reply metrics/src/main.py:66]: yes

File: metrics/src/db.py (hunk 15 above)
Comment (dmytrotkk) [reply metrics/src/db.py]: moved

File: metrics/src/gas.py (hunk 12 above)
Comment (dmytrotkk) [reply metrics/src/gas.py]: fixed

File: metrics/src/main.py (hunk 13 above)
Comment (dmytrotkk) [reply metrics/src/main.py]: moved

**Review by DmytroNazarenko — APPROVED** [review]

**Review by DmytroNazarenko — APPROVED** [review]

**Review by badrogger — APPROVED** [review]

**Review by DmytroNazarenko — APPROVED** [review]

**Pull request merged**